src/d2lfetch-dedupe.js

```diff
@@ -1,51 +1,116 @@
+/**
+ * @typedef {(request: Request) => Promise<Response>} MiddlewareFunc
+ * @typedef {Parameters<ConstructorParameters<typeof Promise>[0]>} PromiseCallbackParams
+ *
+ * @typedef {{resolve: PromiseCallbackParams[0], reject: PromiseCallbackParams[1]}} DeduplicatedRequest
+ *
+ * @typedef {object} InflightRequestEntry
+ * @prop {Request} inflightRequest
+ * @prop {AbortController} [abortController]
+ * @prop {Record<string, DeduplicatedRequest>} dedupedRequests
+ *
+ * @typedef {Record<string, InflightRequestEntry>} InflightRequestInfo
+ */
+
 export class D2LFetchDedupe {
 
 	constructor() {
-		this._inflightRequests = this._inflightRequests || [];
+		/** @type {InflightRequestInfo} */
+		this._inflightRequests = {};
+		this._nextReqId = 0;
 	}
 
+	/**
+	 * @param {Request} request
+	 * @param {MiddlewareFunc} next
+	 *
+	 * @returns {Promise<Response>}
+	 */
 	dedupe(request, next) {
 		if (false === request instanceof Request) {
 			return Promise.reject(new TypeError('Invalid request argument supplied; must be a valid window.Request object.'));
 		}
 
 		if (request.method !== 'GET' && request.method !== 'HEAD' && request.method !== 'OPTIONS') {
-			if (!next) {
```

Comment:
There will always be a `next` middleware (at the very least, the fetch itself), and upstream middleware/consumers wouldn't be expecting a `Request` object